core/drivers/stm32_rng.c

```diff
@@ -367,6 +369,62 @@ TEE_Result hw_get_random_bytes(void *out, size_t size)
 }
 #endif
 
+static TEE_Result stm32_rng_pm_resume(void)
+{
+	vaddr_t rng_base = get_base();
+
+	/* Clean error indications */
+	io_write32(rng_base + RNG_SR, 0);
+
+	if (stm32_rng->ddata->has_cond_reset) {
+		/*
+		 * Correct configuration in bits [29:4] must be set in the same
+		 * access that set RNG_CR_CONDRST bit. Else config setting is
+		 * not taken into account. CONFIGLOCK bit must also be unset but
+		 * it is not handled at the moment.
+		 */
+		io_write32(rng_base + RNG_CR,
+			   stm32_rng->pm_cr | RNG_CR_CONDRST);
+
+		io_clrsetbits32(rng_base + RNG_CR, RNG_CR_CONDRST,
+				RNG_CR_RNGEN);
+	} else {
+		io_write32(rng_base + RNG_CR, RNG_CR_RNGEN | stm32_rng->pm_cr);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_rng_pm_suspend(void)
+{
+	stm32_rng->pm_cr = io_read32(get_base() + RNG_CR);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result
+stm32_rng_pm(enum pm_op op, unsigned int pm_hint __unused,
+	     const struct pm_callback_handle *pm_handle __unused)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(stm32_rng);
+
+	res = clk_enable(stm32_rng->clock);
+	if (res)
+		return res;
+
+	if (op == PM_OP_RESUME)
+		res = stm32_rng_pm_resume();
+	else
+		res = stm32_rng_pm_suspend();
```

Comment:
Look at how you implemented bsec PM (d6df31b03ea5cd8f84b8a9c99f8559f859bc61d3), maybe do the same here:

```c
static TEE_Result stm32_rng_pm(enum pm_op op, unsigned int pm_hint __unused,
			const struct pm_callback_handle *pm_handle __unused)
{
	static uint32_t pm_cr = 0;
	TEE_Result res = TEE_ERROR_GENERIC;

	assert(stm32_rng);

	res = clk_enable(stm32_rng->clock);
	if (res)
		return res;
	
	if (op == PM_OP_SUSPEND)
		pm_cr = io_read32(get_base() + RNG_CR);
	else
		res = stm32_rng_pm_resume(pm_cr);

	clk_disable(stm32_rng->clock);

	return res;
}

Reply:
Ok with your suggestion for consistency.

BTW, I way missing assert (op == PM_OP_SUSPEND || op == PM_OP_RESUME), I'm adding it 👍🏻